templates/Cargo.nix.tera

```diff
@@ -67,7 +67,8 @@ in
     {%- if crate.source.Local.path == "." %}
     src = fetchCrateLocal workspaceSrc;
     {%- else %}
-    src = fetchCrateLocal (workspaceSrc + "/{{ crate.source.Local.path }}");
+    src = fetchCrateLocal workspaceSrc;
```

Comment:
This is causing issues for me, with a mono repo style setup. Our structure looks something like this
```
├── services
│   └── test-service
│       ├── Cargo.toml
│       └── src
│           └── main.rs
└── shared
    └── rust
        ├── Cargo.toml
        ├── bar
        │   ├── Cargo.toml
        │   └── src
        │       └── lib.rs
        └── foo
            ├── Cargo.toml
            └── src
                └── lib.rs
```

We might want to reference `foo` from `test-service`, which would be done by setting the crate path to `../../shared/rust/foo`. With these changes that no longer works, since `workspaceSrc` is `./.`, and so Nix does't know about the relative path.

It seems like the appropriate fix would be to somehow get the per-crate root package, and use that in place of `workspaceSrc`. You would also need to change `workspaceOffset` to be aware of the offset as well. I am not familiar enough with Cargo to know if that info is easily attainable though.

Reply:
Just so I'm clear what needs supporting, there is no top level Cargo.toml right?  shared and services are just directories of a monorepo?

I'm somewhat confused how the previous setup works because the Cargo.nix `packageFun` argument is at most for one workspace right now.  What does the expression look like where `../../shared/rust/foo` is a dependency?

Reply:
The diff for the Cargo.nix would be helpful.  I guess there's more information in `crate.source.Local.path` than I understood.

Reply:
Okay I see several ways that this commit needs improvement

The basic strategy is to build with the full source available but to move any workspace toml out of the way since it will interfere with the synthetic toml created just before building the isolated crate.

Since the relative paths can point quite far over in the source, the workspace Cargo.toml needs to be disabled after the directory change (in mkcrate.nix).

I don't think I tested this well enough.  Thanks for using the branch and catching this issue :+1:

Reply:
> Just so I'm clear what needs supporting, there is no top level Cargo.toml right? shared and services are just directories of a monorepo?

In this instance `shared/rust` is a workspace, but I don't think that has a big impact on the issue. `services` is not a workspace. But the individual services within the `services` folder, are often workspaces. 

> The diff for the Cargo.nix would be helpful. I guess there's more information in crate.source.Local.path than I understood.

Here is an example of the diff. 
```
<     src = fetchCrateLocal workspaceSrc;
<     workspaceOffset = "../../../../shared/rust/fortanix-enclaves";
---
>     src = fetchCrateLocal (workspaceSrc + "/../../../../shared/rust/fortanix-enclaves");
```
 For my `Cargo.nix`files `workspaceSrc` is `./`, and so the relative path was just fetched by Nix like any other path. It seems like cargo2nix isn't aware of the workspace for `shared/rust`, which in my use case is fine since we don't reference any files outside of an individual crate.

Reply:
I had to revert the commit.

The cargo library behavior I tried out is such that I would have to re-implement the workspace discovery to find the workspace for each crate to use it when setting the source.

I was able to make some workarounds for similar situations using the offset.  The idea might have a future for such cases.

A potential wrinkle I uncovered while exploring was that we could have .cargo directories declared in parent directories, and these could interfere with settings while building each crate.